weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts

```diff
@@ -2229,6 +2233,299 @@ export const useProjectStats = (projectId: string) => {
   }, [getTsClient, projectId]);
 };
 
+// Helper function to process first turn inputs
+const processFirstTurnInputs = async (
+  getTsClient: () => traceServerClient.TraceServerClient,
+  projectId: string,
+  groupedCalls: {[turnId: string]: traceServerTypes.TraceCallSchema[]}
+): Promise<Map<string, any>> => {
+  const firstTurnsInputs = new Map();
+  const firstCallsPerTurn = Object.entries(groupedCalls)
+    .map(([turnId, calls]) => {
+      // Build set of all parent IDs within this turn
+      const parentIds = new Set(
+        calls.map(call => call.parent_id).filter(Boolean)
+      );
+
+      // Find leaf calls (calls whose ID is not a parent of any other call)
+      const leafCalls = calls.filter(call => !parentIds.has(call.id));
+
+      // Filter leaf calls for LLM calls (those with usage data)
+      const llmLeafCalls = leafCalls.filter(
+        call => !isEmpty(call.summary?.usage)
+      );
+
+      if (llmLeafCalls.length === 0) {
+        return null;
+      }
+
+      // Sort by started_at ascending to get the first LLM leaf call
+      const sortedLlmLeafCalls = llmLeafCalls.sort((a, b) => {
+        return (
+          new Date(a.started_at).getTime() - new Date(b.started_at).getTime()
+        );
+      });
+
+      return {turnId, call: sortedLlmLeafCalls[0]};
+    })
+    .filter(Boolean);
+
+  // Get inputs for first calls
+  if (firstCallsPerTurn.length > 0) {
+    const callInputs = await getTsClient().callsStreamQuery({
+      project_id: projectId,
+      filter: {call_ids: firstCallsPerTurn.map(item => item!.call.id)},
+      columns: ['inputs', 'turn_id'],
+    });
+
+    (callInputs.calls ?? []).forEach(call => {
+      firstTurnsInputs.set(call.turn_id, call.inputs);
+    });
+  }
+
+  return firstTurnsInputs;
+};
+
+// Helper function to process last turn outputs and statuses
+const processLastTurnData = async (
+  getTsClient: () => traceServerClient.TraceServerClient,
+  projectId: string,
+  groupedCalls: {[turnId: string]: traceServerTypes.TraceCallSchema[]}
+): Promise<{outputs: Map<string, any>; statuses: Map<string, string>}> => {
```

Comment:
`unknown` over `any`. Also, consider plucking off the property from the call object itself.